tests/integration/test_github_readonly.py

```diff
@@ -1,24 +1,34 @@
 import os
+from pathlib import Path
+import unittest
+from ogr.services.github import GithubService
+from requre.online_replacing import record_requests_for_all_methods
+from requre.utils import get_datafile_filename
 
-from ogr import GithubService
-from requre.storage import PersistentObjectStorage
-from requre.utils import StorageMode
-from requre import RequreTestCase
 
-
-class ReadOnly(RequreTestCase):
+@record_requests_for_all_methods()
+class ReadOnly(unittest.TestCase):
     def setUp(self):
         super().setUp()
         self.token = os.environ.get("GITHUB_TOKEN")
-        if PersistentObjectStorage().mode == StorageMode.write and not self.token:
+        if not Path(get_datafile_filename(obj=self)).exists() and not self.token:
```

Comment:
Do we have the test function accessible here? What is the `self` here?

It would be nice to have `cassette` somehow accessible and we now save it as the `._requre_cassette` attribute of the function so this might be the way.

Reply:
self is instance of tests and I uses it as internal knowledge that this object it used for identifying test file if it is pytest, so that it is little bit cryptic.
Without this singleton it raised this issue, that in setup I cannot access cassette, theoretical solution is, similar to moving self.service to property, and just handle mode inside test, not inside setup, then I could use external cassette and look into cassette inside test part and then use your solution via `._requre_cassette` but it is not possible inside setup anyhow.
theoretically there are several ways how to deal with it:
 * do not raise exceptions and just run tests depends on setup tokens, so that then I do not need to deal with some cryptic internal get file names  function
 * your solution, but use it inside test, not setup (because cassette name is not know in setup part)
 * solution with one external cassette and just setup the cassette name inside setup and in test handle itself
 * use the singleton, and some part class for tests what will solve it.

I do not other idea how  we can solve it.

Reply:
> * do not raise exceptions and just run tests depends on setup tokens, so that then I do not need to deal with some cryptic internal get file names  function

Last options I think.

> * your solution, but use it inside test, not setup (because cassette name is not know in setup part)

Doable, but I would try to avoid having the same call in all methods.

> * solution with one external cassette and just setup the cassette name inside setup and in test handle itself
> * use the singleton, and some part class for tests what will solve it.

I would try to avoid having global states completely.

----

It's not so easy to solve. I'll think about that but we can leave it as-is for now.

Reply:
theoretically the easiest way could be, to have that external cassette inside decorators, and in class setUp()
has code like:
```
cassette.storage_file(obj=self)
```
and then have decorator with this `cassette=cassette`

then cassette will not be setup as part of the init, because it exists, so  the code will be just part of setup, but bad is that you then have to cassette=cassette and probably could lead to some sideeffects caused by this shared cassette, similar to singleton usages.